### PR TITLE
Add ClientBuilder::handshake2 for async clients

### PR DESCRIPTION
**Pull request by alexcrichton**
Currently the ClientBuilder::handshake function doesn't work with asynchronous
clients as it just does a loop internally. This function extracts that logic
into a new `MidHandshakeClientBuilder` type and `ClientHandshakeError` enum to
handle the loop externally.

This should allow asynchronous clients to still use `ClientBuilder` and the
`anchor_certificates` function, namely allowing `native-tls` to update to using
`ClientBuilder` and leveraging this logic.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Ah this also adds a few assorted other methods used by native-tls to just export the options of `SslContext` on `ClientBuilder`.

The motivation here is to update native-tls with an `anchor_certificates` function in one form or another!

File: security-framework/src/secure_transport.rs [hunk 1]
```diff
@@ -192,6 +213,82 @@ impl<S> MidHandshakeSslStream<S> {
     }
 }
 
+/// An SSL stream midway through the handshake process.
+#[derive(Debug)]
+pub struct MidHandshakeClientBuilder<S> {
+    stream: Option<MidHandshakeSslStream<S>>,
```
Comment (sfackler) [inline security-framework/src/secure_transport.rs]: Why does this need to be an `Option`?

File: security-framework/src/secure_transport.rs [hunk 2]
```diff
@@ -992,21 +1096,54 @@ impl ClientBuilder {
         self
     }
 
+    /// Use the specified chain of certificates for extra certificates in the
+    /// chain if a custom `identity` is also specified through the `identity`
+    /// method.
+    pub fn identity_chain(&mut self, chain: &[SecCertificate]) -> &mut Self {
```
Comment (sfackler) [inline security-framework/src/secure_transport.rs]: I would probably just call this `certificate_chain`.

**Comment by alexcrichton** [comment]
Updated!

**Comment by alexcrichton** [comment]
Hm ok I pushed another update that should fix the compilation on iOS, but it also raises the question of what to do about protocol configuration on iOS which doesn't have `set_protocol_version_enabled` but does have `set_protocol_version_{min,max}` presumably? Do you think we should infer the min/max?

**Comment by sfackler** [comment]
Might make sense to switch to min/max everywhere and then conditionally compile the implementation.

File: security-framework/src/secure_transport.rs [hunk 3]
```diff
@@ -992,21 +1102,47 @@ impl ClientBuilder {
...
+    pub fn protocol_max(&mut self, max: SslProtocol) -> &mut Self {
+        self.protocol_max = Some(max);
+        self
+    }
+
     /// Initiates a new SSL/TLS session over a stream connected to the specified
     /// domain.
+    ///
+    /// Note that this method assumes that the stream `S` is in *blocking* mode,
+    /// and it will spin unnecessarily if the stream is set to nonblocking mode.
```
Comment (sfackler) [inline security-framework/src/secure_transport.rs]: I think it'll just fail rather than spin?

Comment (alexcrichton) [inline security-framework/src/secure_transport.rs]: Good point! Updated

**Pull request merged**